com.unity.ml-agents/Runtime/Agent.cs

```diff
@@ -801,22 +830,41 @@ void ResetData()
         /// You can also use the [Input System package], which provides a more flexible and
         /// configurable input system.
         /// <code>
-        ///     public override void Heuristic(float[] actionsOut)
+        ///     public override void Heuristic(float[] continuousActionsOut, int[] discreteActionsOut)
         ///     {
-        ///         actionsOut[0] = Input.GetAxis("Horizontal");
-        ///         actionsOut[1] = Input.GetKey(KeyCode.Space) ? 1.0f : 0.0f;
-        ///         actionsOut[2] = Input.GetAxis("Vertical");
+        ///         continuousActions[0] = Input.GetAxis("Horizontal");
+        ///         continuousActions[1] = Input.GetKey(KeyCode.Space) ? 1.0f : 0.0f;
+        ///         continuousActions[2] = Input.GetAxis("Vertical");
         ///     }
         /// </code>
         /// [Input Manager]: https://docs.unity3d.com/Manual/class-InputManager.html
         /// [Input System package]: https://docs.unity3d.com/Packages/com.unity.inputsystem@1.0/manual/index.html
         /// </example>
-        /// <param name="actionsOut">Array for the output actions.</param>
-        /// <seealso cref="OnActionReceived(float[])"/>
-        public virtual void Heuristic(float[] actionsOut)
+        /// <param name="continuousActionsOut">Array to write the continuous actions to.</param>
+        /// <param name="discreteActionsOut">Array to write the discreteActions to.</param>
+        /// <seealso cref="IActionReceiver.OnActionReceived"/>
+        public virtual void Heuristic(float[] continuousActionsOut, int[] discreteActionsOut)
         {
             Debug.LogWarning("Heuristic method called but not implemented. Returning placeholder actions.");
-            Array.Clear(actionsOut, 0, actionsOut.Length);
+            // For backward compatibility
+            switch (m_PolicyFactory.BrainParameters.VectorActionSpaceType)
+            {
+                case SpaceType.Continuous:
+                    #pragma warning disable CS0618
+                    Heuristic(continuousActionsOut);
+                    #pragma warning restore CS0618
+                    Array.Clear(discreteActionsOut, 0, discreteActionsOut.Length);
+                    break;
+                case SpaceType.Discrete:
+                    var convertedOut = Array.ConvertAll(discreteActionsOut, x => (float)x);
```

Comment:
This will allocate twice, right (once here and once for convertedBackToInt)? Maybe better to keep the int[] around and manually convert?

Reply:
actually, I think I can copy back into the float array. For some reason, I just can't copy from float -> int

Reply:
sorry, other way around.